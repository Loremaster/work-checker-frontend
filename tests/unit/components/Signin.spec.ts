import { mount, createLocalVue } from "@vue/test-utils";
import Vuetify from "vuetify";
import Vue from "vue";
import Vuex from "vuex";
import VueRouter from "vue-router";
import flushPromises from "flush-promises";
import Component from "@/components/Signin.vue";

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(VueRouter);

Vue.use(Vuetify);

const router = new VueRouter();

let actions: any;
let store: any;
let $notify: any;
let wrapper: any;

describe("Signin.vue", () => {
  beforeEach(() => {
    actions = { authenticate: jest.fn() };
    $notify = jest.fn();
  });

  const setup = () => {
    store = new Vuex.Store({ actions });
    wrapper = mount(Component, {
      store,
      localVue,
      router,
      mocks: { $notify },
    });
  };

  it("renders correctly", () => {
    setup();

    expect(wrapper).toMatchSnapshot();
  });

  it("submits provided email and password", async () => {
    const email = "user@email.com";
    const password = "password";
    setup();

    await submitWithData({ email, password });

    expect(actions.authenticate).toHaveBeenCalledWith(expect.anything(), { email, password }, undefined);
  });

  it("redirect to root path after submitting credentials", async () => {
    const email = "user@email.com";
    const password = "password";
    setup();
    const spy = jest.spyOn(router, "push");

    await submitWithData({ email, password });

    expect(spy).toHaveBeenCalledWith("/");
    spy.mockClear();
  });

  it("notifies about error", async () => {
    actions.authenticate = jest.fn().mockRejectedValue(new Error("Action error"));
    setup();

    await submit();

    expect($notify).toHaveBeenCalledWith({
      text: "Can't sign in. Please make sure your login and password are correct",
      title: "Authentication",
      type: "error",
    });
  });
});

async function submitWithData({ email, password }: { email: string; password: string }) {
  wrapper.find("input[name='email']").setValue(email);
  wrapper.find("input[name='password']").setValue(password);
  await submit();
}

async function submit() {
  wrapper.find(".submit").trigger("click");
  await flushPromises();
}
