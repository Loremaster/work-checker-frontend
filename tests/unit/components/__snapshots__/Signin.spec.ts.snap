// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signin.vue renders correctly 1`] = `
<div class="elevation-12 v-card v-sheet theme--light">
  <nav class="v-toolbar theme--dark primary" style="margin-top: 0px; padding-right: 0px; padding-left: 0px; transform: translateY(0px);">
    <div class="v-toolbar__content" style="height: 64px;">
      <div class="v-toolbar__title">Login form</div>
      <div class="spacer"></div>
    </div>
  </nav>
  <div class="v-card__text">
    <form novalidate="novalidate" class="v-form">
      <div class="v-input v-text-field theme--light">
        <div class="v-input__prepend-outer">
          <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon material-icons theme--light">person</i></div>
        </div>
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot"><label aria-hidden="true" class="v-label theme--light" style="left: 0px; position: absolute;">Email</label><input aria-label="Email" name="email" type="email"></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <div class="v-messages__wrapper"></div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-input v-text-field theme--light">
        <div class="v-input__prepend-outer">
          <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon material-icons theme--light">lock</i></div>
        </div>
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot"><label aria-hidden="true" class="v-label theme--light" style="left: 0px; position: absolute;">Password</label><input aria-label="Password" name="password" type="password"></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <div class="v-messages__wrapper"></div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
  <div class="v-card__actions">
    <div class="spacer"></div> <button type="button" class="submit v-btn theme--light primary">
      <div class="v-btn__content">Login</div>
    </button>
  </div>
</div>
`;
